src/compiler/builder.ts

```diff
@@ -176,7 +176,7 @@ namespace ts {
         const canCopyEmitSignatures = compilerOptions.composite &&
             oldState?.emitSignatures &&
             !outFilePath &&
-            !compilerOptionsAffectDeclarationPath(compilerOptions, oldCompilerOptions!);
+            !compilerOptionsAffectDeclarationPath(compilerOptions, oldState.compilerOptions);

```

Comment:
I think a comment would help to explain why this one spot doesn't use `oldCompilerOptions`.  (To be honest, I couldn't figure it out from your PR description.)  It's especially disconcerting that we might be dotting into `oldState` when `useOldState` is false, so please consider explaining that as well.

Reply:
Please check if its more clear now.

Reply:
It sounds like maybe `useOldState` should be renamed?  Maybe something like `useOldStateFor...`?

I'll attempt to paraphrase - please check my understanding:
We can only reuse emit signatures (i.e. .d.ts signatures) if the .d.ts file is unchanged, which will only be the case if the `declarationDir` and `outDir` options are unchanged.
We need to look in `oldState.compilerOptions`, rather than `oldCompilerOptions` (i.e. we need to disregard `useOldState`) because `oldCompilerOptions` will be undefined if the file references map indicates that the files to be emitted have changed, which isn't relevant for checking signature reuses.

Was I close?

Reply:
We need to look in oldState.compilerOptions, rather than oldCompilerOptions (i.e. we need to disregard useOldState) it is determined by whether we create reference map to check for files to emit or not. So if say module option changes from `Module.None` to some other `Module` kind then `oldCompilerOptions` will be `undefined` but that shouldn't be taken into account when considering whether to reuse d.ts emit signature

Reply:
> Maybe something like useOldStateFor...? 

not ideal since  `useOldStateFor` since its everything - like semantic diagnostics, files to emit, changed files and so on except d.ts emit signature

Reply:
That's much clearer - thanks!  For my own edification: every use case other than signature emit should still use `oldCompilerOptions`, correct?

Reply:
Yes